### ISSUE TITLE
The computation of weighted CrossEntropyLoss is not reasonable

### ISSUE THREAD

**Issue by hukkai** [pytorch/pytorch#50528]
## 🐛 Bug
Actually this is not a bug in codes, but in the logic.
For torch.nn.CrossEntropyLoss, if `weight` argument is specified, the loss is computed as:
\text{loss} = \frac{\sum^{N}_{i=1} loss(i, class[i])}{\sum^{N}_{i=1} weight[class[i]]}

The normalization term in the denominator is computed given the batch of data.
This will make the results of `reduction=mean` and the results of `reduction=sum`/`reduction=none`  inconsistent. 
And most importantly, if `batch_size=1`, the  `weight` argument  does not work.
### `reduction=mean` and the results of `reduction=sum/none`  inconsistent. 
```
import torch

batch_size, num_class = 8, 10
logits = torch.randn(batch_size, num_class)
target = torch.randint(num_class, size=[batch_size])
ce_loss = torch.nn.CrossEntropyLoss
class_weight = torch.rand(num_class)
ce_loss(weight=class_weight)(logits, target)
ce_loss(weight=class_weight, reduction='none')(logits, target).mean()
```
The two results should be the same, but not the same! Further, the loss of one piece of data will be different in different batches.
### `weight` argument does not work when `batch_size=1`
```
import torch

batch_size, num_class = 1, 10
logits = torch.randn(batch_size, num_class)
target = torch.randint(num_class, size=[batch_size])
ce_loss = torch.nn.CrossEntropyLoss
class_weight = torch.rand(num_class)
ce_loss()(logits, target)
ce_loss(weight=class_weight)(logits, target)
```
The two results should not be the same, but exactly the same!

## Environment

PyTorch version: 1.7.0
Is debug build: True
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 11.0.1 (x86_64)
GCC version: Could not collect
Clang version: 12.0.0 (clang-1200.0.32.27)
CMake version: Could not collect

Python version: 3.8 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.4
[pip3] torch==1.7.0
[pip3] torchvision==0.8.1
[conda] numpy                     1.19.4                   pypi_0    pypi
[conda] torch                     1.7.0                    pypi_0    pypi
[conda] torchvision               0.8.1                    pypi_0    pypi

cc @albanD @mruberry @jbschlosser

**Comment by jonykarki** [comment]
I would like to work on this if it's not super difficult.

**Comment by mruberry** [comment]
@jonykarki We should let the team review and make a statement about (1) whether they would accept a fix and (2) what kind of fix would be acceptable before proceeding to a PR.

If you're looking for an issue, however, maybe I can help you find one if you let me know what you're interested in?

**Comment by jonykarki** [comment]
@mruberry 
Nothing in particular. Anything that's approachable will work.

**Comment by hukkai** [comment]
@mruberry @jonykarki I think the coding is not difficult but I cannot find the codes. As I see it, this function is called from torch/_C.cpython-38-darwin.so

**Comment by mruberry** [comment]
> Nothing in particular. Anything that's approachable will work.

This issue just came in recently: https://github.com/pytorch/pytorch/issues/50790. Does that sound interesting, @jonykarki?

**Comment by jonykarki** [comment]
Yes. I can work on it. Thank you.
